### PR TITLE
feat(sourcemaps): Add more indexes to tables for artifact bundles

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR updates indexes on the artifact bundles tables to account for our access patterns, removes uniqueness constraints that won't be supported by partitioning and removes some indexes that were not used.

### CODE REVIEW

File: src/sentry/models/artifactbundle.py
Comment (wedamija) [inline src/sentry/models/artifactbundle.py:157]: What sorts of queries do you need to make here? Will you be querying on `organization_id, dist_name` without a `release_name` filter at all?

Comment (iambriccardo) [reply src/sentry/models/artifactbundle.py:157]: The reasoning for this is that we have a lot of equal `release/dist` pairs across orgs, thus we would benefit from scoping down the org directly in the index.

Comment (iambriccardo) [reply src/sentry/models/artifactbundle.py:157]: How we would query is (simplified):
```sql
SELECT *
FROM sentry_releaseartifactbundle AS t1 JOIN sentry_projectartifactbundle AS t2 JOIN sentry_artifactbundle AS t3
WHERE t1.organization_id = x AND t1. release_name = y AND t1.dist_name = z AND t2.project_id = a
```

In addition, the reasoning for including the artifact bundle in the composite index is to leverage a single index for both joining and filtering since in other queries we assume the dbms will choose to first filter out `ArtifactBundle` and then join it on the other tables with the `artifact_bundle_id` and the other fields specific to those tables (e.g., `release_name`, `url`, `debug_id`).

Comment (Swatinem) [reply src/sentry/models/artifactbundle.py:157]: In particular, I updated some of the queries explicitly to use that specific index:
https://github.com/getsentry/sentry/pull/52611

Initially, the queries performed a lot worse, but I figured out there was a skew that we were disproportionately querying orgs/releases with a ton of bundles.

Now that this is fixed, the optimized query performs better, clocking in at around ⅔ the *average* time as the query without using the composite index.

File: src/sentry/models/artifactbundle.py
Comment (Swatinem) [inline src/sentry/models/artifactbundle.py:180]: For all the cases where we replace a `unique` constraint with a simple index:

NOTE to other reviewers: We are doing this because table partitioning (which we eventually want to do for quick deletes) is incompatible with unique constraints :-(

However, we should be extra careful with all of these to ensure on the writing side that we are not creating duplicates. As those are not enforced directly on database level, we can never completely remove the likelihood though :-(

**Review by Swatinem — APPROVED** [review]

File: src/sentry/models/artifactbundle.py
Comment (wedamija) [reply src/sentry/models/artifactbundle.py:180]: You might want try use a distributed lock on the write path to help avoid double writes. That'd probably guarantee it - you could just hash the unique attributes together and take out a lock when writing. Take a look around the codebase for `lock.acquire()` for some examples.

File: src/sentry/models/artifactbundle.py
Comment (wedamija) [reply src/sentry/models/artifactbundle.py:157]: Sounds good, just wanted to make sure you never needed to query on `release_name, artifact_bundle` etc, since that'd be a different index

**Review by wedamija — APPROVED** [review]

**Review by wedamija — DISMISSED** [review]
Just realised this needs to be `is_dangerous=True`, since `sentry_debugidartifactbundle` has 124mm rows. The migration won't finish in the deploy cycle.

Feel free to dismiss this review once you've updated so that you can merge

**Review by beezz — APPROVED** [review]

**Pull request merged**
